songpal/method.py

```diff
@@ -152,18 +155,63 @@ async def __call__(self, *args, **kwargs):
 
     @property
     def inputs(self) -> Dict[str, type]:
-        """Input parameters for this method."""
-        return self.signature.input
+        """Input parameters for this method version."""
+        return self.signatures[self._version].input or {}
+
+    @property
+    def latest_supported_version(self) -> Optional[str]:
+        """Latest version supported by this method."""
+        return (
+            max(self._supported_versions) if len(self._supported_versions) > 0 else None
+        )
 
     @property
     def outputs(self) -> Dict[str, type]:
-        """Output parameters for this method."""
-        return self.signature.output
+        """Output parameters for this method version."""
+        return self.signatures[self._version].output or {}
+
+    @property
+    def supported_versions(self) -> Set[str]:
+        """List of supported version numbers for this method."""
+        return self._supported_versions
+
+    @property
+    def version(self) -> str:
+        """Method signature version number."""
+        return self._version
+
+    def add_supported_version(self, version: str):
+        """Add a supported version number for this method."""
+        self._supported_versions.add(version)
+
+    def supports_version(self, version: str) -> bool:
+        """Is this method version supported."""
+        return version in self._supported_versions
+
+    def use_version(self, version: str):
+        """Specify method signature version to use."""
+        if self.supports_version(version):
+            if version in self.signatures:
+                self._version = version
+            else:
+                raise SongpalException(
+                    "Version %s of %s.%s is supported, but no signature is available"
+                    % (version, self.service.name, self.name)
+                )
```

Comment:
This reads error-prone to me, but I'm afraid I don't have a nice way to solve it right now. I suppose it is as it is, and we fix it if deemed necessary :-)

Reply:
Actually this checking is only needed if return a bool and use it like below, which I haven't actually done.
```
if method.use_version("1.2"):
    res = await method({})
else:
    res = await method()
```
otherwise it is already being checked
```
if method.supports_version("1.2"):
    method.use_version("1.2")
    res = await method({})
else:
    res = await method()
```
The signature check isn't needed as the param type checking isn't being used. So even with the first example it can be reduced. What would be your preference? Opt. 1 may need to rename `use_version()`

Reply:
Considering that we are changing the internal state, it's better to be explicit about it, so the second one.

Reply:
Cool, then none of the checks are currently necessary.